force-app/main/default/classes/SOQL.cls

```diff
@@ -270,92 +356,182 @@ public with sharing class SOQL {
      * @return List of contacts with the same Mailing State as the Account Billing State
      */
     public static List<Contact> getAccountsWithContactsByState(Id accountId) {
+        List<Contact> matchingContacts = new List<Contact>();
         // Query the Account
-
+        Account acc = [SELECT BillingState FROM Account WHERE Id = :accountId LIMIT 1];
+       
         // Store the Account's BillingState
+        String billingState = acc.BillingState;
 
         // Query the Contacts
-        return null; // Replace null with the variable you used to store the result
+        matchingContacts = [SELECT Id, FirstName, LastName, MailingState 
+                            FROM Contact 
+                            WHERE MailingState = :billingState];
+        return matchingContacts; // Replace null with the variable you used to store the result
     }
 
-    /**
-     * Question 22
-     * Extracts email domains from them leads, then finds Accounts with matching website domains
-     * If Lead Email = test@test.com return account with website = www.test.com
-     * Example: findAccountsByLeadEmailDomains([set of lead ids]) should return a list of Accounts with matching email domains
-     * @param A set of Lead Ids.
-     * @return A list of Accounts with matching email domains.
-     */
-    public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
-        // Query the Leads
-
-        // Create a set of email domains from the Leads
-
-        // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
+/**
+* Question 22
+* Extracts email domains from them leads, then finds Accounts with matching website domains
+* If Lead Email = test@test.com return account with website = www.test.com
+* Example: findAccountsByLeadEmailDomains([set of lead ids]) should return a list of Accounts with matching email domains
+* @param A set of Lead Ids.
+* @return A list of Accounts with matching email domains.
+*/
+public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
         
-        // Query the Accounts
-        return null; // Replace null with the variable you used to store the result
+    Set<String> emailDomains = new Set<String>();
+    for (Lead lead : [SELECT Id,company, Email FROM Lead WHERE Id IN:leadIds]) 
+    {
+        // Get the email from the Lead
+        String email = lead.Email;
+        
+        // Check if the email is not null
+        if (email != null) 
+        {
+            //system.debug('this is email when not null => '+email);//this is email when not null => bertha@fcof.net
+            String emailDomain = email.split('@') [1];//this is emailDomain from email => fcof.net
+            //system.debug('this is emailDomain from email => '+emailDomain);
+            
+            // Find the last dot in the domain to remove the TLD
```

Comment:
In this case, it probably wasn't necessary to remove the TLD. Essentially, we're trying to convert an email (warren@**cloudcodeacademy.com**) to it's domain name (www **cloudcodeacademy.com**). But nice job thinking it through!